[Encryption] Add support for Field-Level Automatic and Queryable Encryption

GromNaN opened the pull request:
|      Q       |   A
|------------- | -----------
| Type         | feature
| BC Break     | no
| Fixed issues | [PHPORM-317](https://jira.mongodb.org/browse/PHPORM-317)

#### Summary

⚠️ This feature requires a MongoDB Atlas or Enterprise license.

> [Client-Side Field Level Encryption (CSFLE)](https://www.mongodb.com/docs/manual/core/csfle/) is a feature that enables you to encrypt data in your application before you send it over the network to MongoDB. With CSFLE enabled, no MongoDB product has access to your data in an unencrypted form.

> [Queryable Encryption](https://www.mongodb.com/docs/manual/core/queryable-encryption/) gives you the ability to perform the following tasks:
> - Encrypt sensitive data fields from the client-side.
> - Store sensitive data fields as fully randomized encrypted data on the database server-side.
> - Run expressive queries on the encrypted data.

Implementation:
- Add `#[Encrypt]` attribute with [field encryption options](https://www.mongodb.com/docs/manual/core/queryable-encryption/fundamentals/encrypt-and-query/#configure-encrypted-fields-for-optimal-search-and-storage).
- Add `autoEncryption` configuration with [options to create an encryption client](https://www.php.net/mongodb-driver-clientencryption.construct) (MongoDB collection that stores encryption keys for each field)

The encrypted collection must be created using the schema manager (`odm:schema:create` command)

GromNaN committed:
CSFLEQE

GromNaN committed:
Add test document classes from tutorial

https://github.com/mongodb/docs/blob/master/source/includes/qe-tutorials/node/queryable-encryption-tutorial.js

GromNaN committed:
Create encryption field map

GromNaN committed:
Create encrypted collection

GromNaN committed:
Use promoted properties for encrypted document tests

GromNaN committed:
Revert changes in BaseTestCase

GromNaN committed:
Rename EncryptionFieldMapTest

GromNaN committed:
Create Configuration::getDriverOptions() to create the client

GromNaN committed:
Support class-level #[Encrypt] attribute

GromNaN committed:
Skip QE tests on non-supported server configuration

GromNaN committed:
Add documentation on #[Encrypt] attribute

GromNaN committed:
Add XML mapping for "encrypt" tag

GromNaN committed:
Create specific xsd type for embedded documents to enable <encrypt> tag

GromNaN committed:
Fix import class Throwable

GromNaN committed:
Fix access to $version property before initialization

alcaeus commented on docs/en/reference/attributes-reference.rst:
```diff
@@ -336,6 +336,40 @@ Unlike normal documents, embedded documents cannot specify their own database or
+#[Encrypt]
+----------
+
+The `#[Encrypt]` attribute is used to define an encrypted field mapping for a document property. It allows you to configure fields for encryption and queryable encryption in MongoDB.
```
RST syntax requires two backticks for inline code:

```suggestion
The ``#[Encrypt]`` attribute is used to define an encrypted field mapping for a document property. It allows you to configure fields for encryption and queryable encryption in MongoDB.
```

alcaeus commented on docs/en/reference/attributes-reference.rst:
```diff
@@ -336,6 +336,40 @@ Unlike normal documents, embedded documents cannot specify their own database or
+        public string $name;
+    }
+
+For more details, refer to the MongoDB documentation on [Queryable Encryption](https://www.mongodb.com/docs/manual/core/queryable-encryption/fundamentals/encrypt-and-query/).
```
Wrong link syntax:

```suggestion
For more details, refer to the MongoDB documentation on  `Queryable Encryption <https://www.mongodb.com/docs/manual/core/queryable-encryption/fundamentals/encrypt-and-query/>`_.
```

alcaeus commented on doctrine-mongo-mapping.xsd:
```diff
@@ -16,7 +16,7 @@
       <xs:choice minOccurs="0" maxOccurs="unbounded">
         <xs:element name="document" type="odm:document" minOccurs="0" maxOccurs="unbounded" />
-        <xs:element name="embedded-document" type="odm:document" minOccurs="0" maxOccurs="unbounded" />
+        <xs:element name="embedded-document" type="odm:embedded-document" minOccurs="0" maxOccurs="unbounded" />
```
Noted that this change will disappear when #2766 is merged.

GromNaN replied:
I rebased the PR.

alcaeus commented on lib/Doctrine/ODM/MongoDB/DocumentManager.php:
```diff
@@ -225,6 +220,21 @@ public function getClient(): Client
         return $this->client;
     }
 
+    public function getClientEncryption(): ClientEncryption
```
We may want to mark this method as internal:
```suggestion
    /** @internal **/
    public function getClientEncryption(): ClientEncryption
```

alcaeus commented on lib/Doctrine/ODM/MongoDB/Mapping/Annotations/Encrypt.php:
```diff
@@ -0,0 +1,49 @@
+     * @param positive-int|null       $contention
+     */
+    public function __construct(
+        public ?string $queryType = null,
```
Would it make sense to use a backed enum for this?

GromNaN replied:
I created the `EncryptQuery` enum in the same namespace. So that it can be used like this:
```php
#[ODM\Encrypt(queryType: ODM\EncryptQuery::Equality)]
```

alcaeus commented on doctrine-mongo-mapping.xsd:
```diff
@@ -26,7 +26,38 @@
+      <xs:element name="field" type="odm:field" minOccurs="0" maxOccurs="unbounded" />
+      <xs:element name="embed-one" type="odm:embed-one" minOccurs="0" maxOccurs="unbounded" />
+      <xs:element name="embed-many" type="odm:embed-many" minOccurs="0" maxOccurs="unbounded" />
+      <xs:element name="encrypt" type="odm:encrypt-field" minOccurs="0" maxOccurs="1" />
```
Why is this an element instead of an attribute? Is it to future-proof this in case there will be options for encrypting entire documents in future?

GromNaN replied:
Yes, I plan to add a "key" attribute to the element.

GromNaN replied:
As discussed, the "key" don't need to be configured with queryable encryption. I keep the distinct XML attribute for extensibility.

alcaeus commented on lib/Doctrine/ODM/MongoDB/SchemaManager.php:
```diff
@@ -643,10 +644,29 @@ public function createDocumentCollection(string $documentName, ?int $maxTimeMs =
+                $this->dm->getClientEncryption(),
+                $this->dm->getConfiguration()->getKmsProvider(),
+                null, // @todo when is it necessary to set the master key?
+                $options,
```
Should you be using `$this->getWriteOptions` here like below and add the encrypted field config?

alcaeus reviewed (commented):
Left some questions, and phpstan seems to be unhappy about some of the mapping types.

The logic and tests for it look solid.

GromNaN commented on tests/Doctrine/ODM/MongoDB/Tests/Functional/QueryableEncryptionTest.php:
```diff
@@ -0,0 +1,89 @@
+
+    public function testCreateAndQueryEncryptedCollection(): void
+    {
+        // @todo skip if not using MongoDB < 7, single node or not enterprise
```
Done.

GromNaN committed:
Make XSD more strict for other Document types (#2766)

GromNaN committed:
Use an enum for query type

GromNaN committed:
Make getClientEncryption internal

GromNaN committed:
Improve type of min/max bounds for range queries

GromNaN commented on lib/Doctrine/ODM/MongoDB/Mapping/Driver/XmlDriver.php:
```diff
@@ -307,6 +314,23 @@ public function loadMetadataForClass($className, \Doctrine\Persistence\Mapping\C
+                    foreach ($field->encrypt->attributes() as $encryptKey => $encryptValue) {
+                        $mapping['encrypt'][$encryptKey] = match ($encryptKey) {
+                            'queryType' => (string) $encryptValue,
+                            'min', 'max' => match ($mapping['type']) {
+                                Type::INT => (int) $encryptValue,
+                                Type::FLOAT => (float) $encryptValue,
+                                Type::DECIMAL128 => new Decimal128((string) $encryptValue),
+                                Type::DATE, Type::DATE_IMMUTABLE => new UTCDateTime(new DateTimeImmutable((string) $encryptValue)),
+                                default => null, // Invalid
+                            },
```
I've added the transformation of values from string to native BSON types here; should use the `Type::convertToDatabaseValue()` inside `EncryptionFieldMap`.

GromNaN committed:
Use getWriteOptions with createEncryptedCollection

GromNaN committed:
Use random local master key

GromNaN committed:
Add assertion on non-decrypted data

GromNaN committed:
Catch BulkWriteException, since WriteException is removed from ext-mongodb 2.0 (#2770)

GromNaN committed:
Ignore all DOCTRINE_MONGODB_DATABASE phpstan errors

GromNaN committed:
Merge release 2.11.2 into 2.12.x (#2772)

GromNaN committed:
Baseline phpstan

GromNaN committed:
Deprecate APM\Command::getServer(), add getPort and getHost instead (#2773)

CommandSucceededEvent|CommandFailedEvent::getServer() has been deprecated in ext-mongodb 1.20 and removed in 2.0.0.
For backward compatibility: we keep this method in MongoDB ODM, but throw an exception the extension is not compatible. I'm not adding a trigger_deprecation as this is already triggered by the extension method call.

The new methods getPort() and getHost() are added as proxy to extension methods that are always available in the required ext-mongodb 1.21+.

This is not covered by tests, I can add tests on CommandLogger before merging.

GromNaN committed:
Remove unnecessary `assert` calls (#2775)

* CursorInterface always extends Traversable, no need to assert on SPL Iterator
* Remove duplicate assert

GromNaN committed:
Fix CS and skip phpstan issues

alcaeus commented on lib/Doctrine/ODM/MongoDB/SchemaManager.php:
```diff
@@ -643,10 +644,29 @@ public function createDocumentCollection(string $documentName, ?int $maxTimeMs =
+                $class->getCollection(),
+                $this->dm->getClientEncryption(),
+                $this->dm->getConfiguration()->getKmsProvider(),
+                null, // @todo when is it necessary to set the master key?
```
To leave a quick answer, some KMS providers require a master key (e.g. AWS)

alcaeus commented on tests/Doctrine/ODM/MongoDB/Tests/BaseTestCase.php:
```diff
@@ -192,6 +193,24 @@ protected function skipTestIfSharded(string $className): void
+            new Command(['buildInfo' => 1]),
+        )->toArray()[0];
+
+        if (! in_array('enterprise', $buildInfo->modules ?? [])) {
```
I noticed that atlas-local does not expose an enterprise module, so we may want to remove this check before wrapping up the feature.
